File: leetcode-solutions/563-binary-tree-tilt.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
int findDiffDFS(TreeNode* node, int& tilt) {

    if (node == NULL) {
        return 0;
    }

    int l = findDiffDFS(node->left, tilt);
    int r = findDiffDFS(node->right, tilt);

    tilt += abs(l - r);

    return l + r + node->val;
}

int findTilt(TreeNode* root) {

    int tilt = 0;
    findDiffDFS(root, tilt);
    return tilt;
}

File: leetcode-solutions/563-binary-tree-tilt.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var findTilt = function (root) {
    let tilt = 0;

    function findDiffDFS(node) {

        if (!node) {
            return 0;
        }

        let l = findDiffDFS(node.left);
        let r = findDiffDFS(node.right);

        tilt += Math.abs(l - r);

        return l + r + node.val;
    }

    findDiffDFS(root);

    return tilt;
};
